include/golioth/ota.h

```diff
@@ -68,6 +68,14 @@ enum golioth_ota_reason
     GOLIOTH_OTA_REASON_FIRMWARE_UPDATE_FAILED,
     /// Protocol not supported
     GOLIOTH_OTA_REASON_UNSUPPORTED_PROTOCOL,
+    /// IO error while trying to store component
+    GOLIOTH_OTA_REASON_IO,
```

Comment:
@szczys we also have the `NOT_ENOUGH_FLASH_MEMORY` and `OUT_OF_RAM` reasons. It looks like there is [one use of the former](https://github.com/golioth/golioth-firmware-sdk/blob/5434b10fda4321b889478cde2da0fc9f5059a432/src/fw_update.c#L369) and none of the latter. Do you anticipate the being used in different scenarios or eventually replacing those?

Reply:
I discussed this with @sam-golioth and he suggested adding the new code without removing the others as these are available for users to use in their own implementations.

Reply:
Just to confirm, the reason we would introduce this less specific reason is that it allows us to report when we aren't sure exactly where data is being written?

Reply:
The use of `GOLIOTH_OTA_REASON_NOT_ENOUGH_FLASH_MEMORY` was added recently, before that we weren't specifying any reason other than `GOLIOTH_OTA_REASON_FIRMWARE_UPDATE_FAILED`. In it's current usage, we're translating from `GOLIOTH_ERR_IO` (an internal error code) to `GOLIOTH_OTA_REASON_NOT_ENOUGH_FLASH_MEMORY` which makes assumptions about the implementation of the port layer that we probably shouldn't. We should either continue to report `GOLIOTH_OTA_REASON_FIRMWARE_UPDATE_FAILED` or use the new `GOLIOTH_OTA_REASON_IO` code.

Part of the confusion here is that these error codes (improperly, imo) mix SDK errors with errors returned from user callbacks (or the port layer in the case where the fw_update module is the user of the OTA service, but the port layer can also be customized and overridden by customers). The SDK errors are under our control, we know the failure modes, and we can log them precisely. The error codes that indicate issues in user callbacks (specifically `OUT_OF_FLASH` and `OUT_OF_RAM`) are speculative - we're trying to predict what errors a user supplied component may return, and then use these error codes to provide information to the user to help them debug their own code. Fundamentally I think that's a losing game, and we should instead be using the OTA reason codes to indicate that an error occurred in the user callback while providing a flexible mechanism that allows users to provide more information about that error in a way that is useful for their code. In my opinion, that flexible mechanism is either going to be the Golioth logging service or a third-party observability platform through pipelines. Limiting users to pre-defined error codes is always going to cause some friction and confusion trying to get square pegs through round holes.

The main reason for keeping existing error codes around is to not mess up the enum numbering and break compatibility with previous versions of the SDK. We could rename the enums in the SDK to something like `GOLIOTH_OTA_REASON_DEPRECATED`, but as long as we have to keep the values around anyway we might as well make them available for use.

Reply:
> Fundamentally I think that's a losing game, and we should instead be using the OTA reason codes to indicate that an error occurred in the user callback while providing a flexible mechanism that allows users to provide more information about that error in a way that is useful for their code. In my opinion, that flexible mechanism is either going to be the Golioth logging service or a third-party observability platform through pipelines. Limiting users to pre-defined error codes is always going to cause some friction and confusion trying to get square pegs through round holes.

This is the basis of my concerns around expansion of states / reasons here. I think we should either have a very narrow state machine where we are effectively saying "use this if you want to take advantage of some of the automatic insights Golioth's OTA service offers", otherwise emit data to one of the more flexible services you mention. I'd almost go so far to say that the existing reasons we have should be managed through a more flexible service. They feel more like a second level observability indicator ("why is it happening") rather than a primary ("what is happening"). It is also a little strange that any permutation of state and reason can be submitted, when clearly some of the reasons are not applicable to all states.

As far as the move to using the `GOLIOTH_OTA_REASON_IO`, that makes sense to me and I appreciate the attention to backwards compatibility. I do think that this list could continue to grow without a clear definition of the level of specificity we are looking for though, so I think it makes sense to explore whether we can find a better balance of combining a rigid happy path with flexibility that allows for users to determine their own required level of specificity.